plugin/backend_linux.go

```diff
@@ -797,6 +808,251 @@ func (pm *Manager) CreateFromContext(ctx context.Context, tarCtx io.ReadCloser,
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpDir)
+
+	if err := chrootarchive.Untar(input, tmpDir, nil); err != nil {
+		return err
+	}
+
+	f, err := os.Open(filepath.Join(tmpDir, "index.json"))
+	if err != nil {
+		return errors.Wrap(err, "error opening oci index")
+	}
+	defer f.Close()
+	var index ocispecv1.Index
+
+	if err := json.NewDecoder(f).Decode(&index); err != nil {
+		return errors.Wrap(err, "error reading oci index")
+	}
+
+	for _, m := range index.Manifests {
+		if m.MediaType != ocispecv1.MediaTypeImageManifest {
+			continue
+		}
+
+		f, err := os.Open(filepath.Join(tmpDir, "blobs", m.Digest.Algorithm().String(), m.Digest.Hex()))
+		if err != nil {
+			return errors.Wrap(err, "error opening manifest")
+		}
+
+		var manifest ocispecv1.Manifest
+		if err := json.NewDecoder(f).Decode(&manifest); err != nil {
+			return errors.Wrap(err, "error reading manifest")
+		}
+		if manifest.Config.MediaType != schema2.MediaTypePluginConfig {
+			continue
+		}
+
+		// TODO(@cpuguy83): verify digests?
+		f, err = os.Open(filepath.Join(tmpDir, "blobs", manifest.Config.Digest.Algorithm().String(), manifest.Config.Digest.Hex()))
+		if err != nil {
+			return errors.Wrap(err, "error opening plugin config")
+		}
+		defer f.Close()
+
+		configBlob, err := pm.blobStore.New()
+		if err != nil {
+			return err
+		}
+		defer configBlob.Close()
+		r := io.TeeReader(f, configBlob)
+
+		var config types.PluginConfig
+		if err := json.NewDecoder(r).Decode(&config); err != nil {
+			return errors.Wrap(err, "error reading plugin config")
+		}
+		if _, err := configBlob.Commit(); err != nil {
+			return err
+		}
+		privileges, err := computePrivileges(config)
+		if err != nil {
+			return err
+		}
+
+		// create a tmpdir to extract the plugin rootfs
+		tmpRootFS, err := ioutil.TempDir(tmpDir, ".rootfs")
+		if err != nil {
+			return errors.Wrap(err, "error creating rootfs dir")
+		}
+		defer os.RemoveAll(tmpRootFS)
+
+		layerDigests := make([]digest.Digest, 0, len(manifest.Layers))
+		for _, l := range manifest.Layers {
+			if l.MediaType != ocispecv1.MediaTypeImageLayerGzip {
+				continue
+			}
+			compressedTar, err := os.Open(filepath.Join(tmpDir, "blobs", l.Digest.Algorithm().String(), l.Digest.Hex()))
+			if err != nil {
+				return err
+			}
+			defer compressedTar.Close()
+			// XXX: Instead of using the blobStore interface to create, write and commit
+			// to blobstore, why not simply copy the files and rename them?
+			rootfsBlob, err := pm.blobStore.New()
+			if err != nil {
+				return err
+			}
+			defer rootfsBlob.Close()
+
+			size, err := io.Copy(rootfsBlob, compressedTar)
+			if err != nil {
+				return errors.Wrap(err, "error reading layer")
+			}
+			if size != l.Size {
+				return errors.New("layer size mismatch")
+			}
+
+			if _, err := rootfsBlob.Commit(); err != nil {
+				return err
+			}
+
+			// decompress for plugin rootfs
+			if _, err := compressedTar.Seek(0, 0); err != nil {
```

Comment:
This can be done with `TeeReader` instead of reading twice. Look at the pull code, or even better to extract this into same function that both pull and load use https://github.com/moby/moby/blob/master/plugin/blobstore.go#L141-L153

Should also verify the blob against the config. https://github.com/moby/moby/blob/master/plugin/blobstore.go#L149